### PR TITLE
[JIT] Separates CPU and CUDA fusers

### PR DESCRIPTION
**Pull request by mruberry**
This PR separates the CPU and CUDA fusion passes and compilers, created a new fusers subdirectory to store the logic for each. The passes were updated to only operate on CPU or CUDA nodes, and the compilers have checks to ensure they're dealing with the correct type. More precisely:

-Creates a new "fusers" directory
-The graph_fuser pass now calls a cpu fusion pass if we can compile on the cpu
-The graph_fuser pass also calls a cuda fusion pass if we're using cuda, not ROCm, and not Windows
-The cuda fuser is only built if using cuda, not ROCm, and not Windows
-The passes now operate only on nodes of their respective types
-The fusers now operate only on nodes of their respective types
-Both fusers have some clean-up refactoring, mostly spinning classes off into their own headers

There is no end-user functionality change with this PR, but it sets the stage for further fusion enhancements.

Note: this PR was spun out of #10544, since it contains additional improvements. Originally it was to diagnose a cpu-only trusty failure in the latter (hence the name "cuda_fuser_trusty"). Thanks to @ezyang for being incredibly helpful in diagnosing that issue. Closing #10544 in favor of this branch.

@csarofeen also contributed to this PR, as can be seen in the history of #10544, which I plundered.

### CODE REVIEW

**Comment by mruberry** [comment]
The failure on caffe2-conda2-macos10.13 appears unrelated and that build was passing previously.

File: torch/csrc/jit/fusers/cpu/annotated_graph.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+TORCH_API struct AnnotatedGraph {
+  // short-term storage only, so it borrows Graph.
+  AnnotatedGraph(Graph& graph, int device)
+  : graph(&graph), device(device) {}
+
+  Graph* graph = nullptr;
+  int device = kCPUDevice;
+  std::vector<TensorDesc> input_desc;
+  std::vector<TensorDesc> output_desc;
+};
```
Comment (apaszke) [inline torch/csrc/jit/fusers/cpu/annotated_graph.h]: Is this really CPU specific in any way? It looks like this one, `TensorDesc` and a few other structs are shared and we're unnecessarily duplicating code. Can you move them as a shared util to the `fusers` directory?

File: torch/csrc/jit/fusers/cuda/simple_ops.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+namespace torch { namespace jit {
+
+// What is a simple mappable operator?  It is:
+//    - Has an output with the same sizes as its input
+//    - Single output
+//    - Can handle non-contiguous input
+//    - Produces contiguous output
+// Some of these restrictions may be relaxable, but you should
+// carefully read the code first, as we rely on these assumptions.
+std::unordered_set<NodeKind> simple_mappable = {
```
Comment (apaszke) [inline torch/csrc/jit/fusers/cuda/simple_ops.h]: This is a non-inline non-static map in a header. That's not a good idea. This is really needed only in the fuser pass, so I don't understand why did this end up in a header

**Review by apaszke — CHANGES_REQUESTED** [review]
Thanks @mruberry, that's a step in a good direction, but I see a few problems:
1. There's *a lot* of code duplication. If you look at the +- lines of this diff, it ends up pretty much doubling the size of our fuser. We should resolve this before we land this PR. Some things should really be trivial to share like `AnnotatedGraph` or `TensorDesc`. Also, GraphFusers for CPU and CUDA aren't all that different, and so we might want to have a common base implementation for them (with a few virtual hooks to customize what they do). Finally, the codegen is also very similar between CPU and CUDA. Only the syntax at the beginning and the end of the generated function changes.
2. The fuser is pretty broken in multiple ways at the moment, because of the recent refactors in our codebase. @zou3519 have sent a few PRs that fix those issues, but they will heavily conflict with this PR. We need to coordinate those. Finally, touching upon 1. again, if we duplicate the code, we'll need to fix all those problems twice.
3. I don't feel like splitting tiny helper data structures like AnnotatedGraph into separate headers (unless they are really shared by a lot of files) is beneficial. I honestly hate having a thousand files open in my editor, and I find that it's generally better to divide things based on their semantics and similar functionality, instead of the "one class per file only" rule.

**Comment by soumith** [comment]
> Finally, the codegen is also very similar between CPU and CUDA. Only the syntax at the beginning and the end of the generated function changes.

@apaszke you should consider that this is the point of divergence. A couple of months in, the CPU and CUDA fusers will look quite different

File: torch/csrc/jit/fusers/cuda/simple_ops.h (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/fusers/cuda/simple_ops.h]: It has everything to do with the logic of the file it was inside of. It tells you which nodes are eligible for fusion, and in fact it doesn't even account for all of them, which might have more complicated rules. I don't see any reason to lift it out.

Comment (csarofeen) [reply torch/csrc/jit/fusers/cuda/simple_ops.h]: The only thing the pass cares about is `bool isSimpleMap(Node* node)` for each node that's passed. It doesn't care about what the function is just if it is a simple map or not.
However, if you like it better the other way I'm happy if we put it back in. Don't think this needs to be a sticking point.

**Comment by mruberry** [comment]
Closing to wait for @apaszke to improve separation of fusion pass and fusion compiler. Plan is for fusion pass to be outside the fuser. Huzzah! Also, we will create a "directory" for code overlap between the cpu and CUDA fusers in the updated PR. Though the fusers may diverge later, while they share common code it may as well actually be shared. 

fyi @ngimel

**Pull request closed without merging**